### DISCUSSION TITLE
How do I contribute to the project?

### DISCUSSION THREAD

**Discussion by garydwatson** [vim/vim#13087] [Ideas]
I'm interested in contributing to the project...  Starting up this discussion thread for two reasons...

  1. To find out the best way I can personally contribute
  2. Other people might be interested in contributing too, if so, whatever answers I get, they might find useful too :)

To that end...
  1. What tasks could I help with (assuming something from the issue tracker might be good)
  2. Are there any conventions I should know about to not get made fun of too much :)
  3. Where's the best place to communicate about such things (or is this discussion section it?)
    1. side note, is there a good live chat somewhere?

**Comment by vim-ml** [comment]
Hi,

On Thu, Sep 14, 2023 at 10:23 AM Gary Watson ***@***.***>
wrote:

> I'm interested in contributing to the project... Starting up this
> discussion thread for two reasons...
>
>    1. To find out the best way I can personally contribute
>    2. Other people might be interested in contributing too, if so,
>    whatever answers I get, they might find useful too :)
>
> To that end...
>
>    1. What tasks could I help with (assuming something from the issue
>    tracker might be good)
>    2. Are there any conventions I should know about to not get made fun
>    of too much :)
>    3. Where's the best place to communicate about such things (or is this
>    discussion section it?)
>    4. side note, is there a good live chat somewhere?
>

Thanks for reaching out. I would recommend the following tasks:

1. Develop test scripts for the Vim features.  There are a large number of example scripts in the [src/testdir directory](https://github.com/vim/vim/tree/master/src/testdir).  Vim has an extensive test infrastructure for developing different types of tests.  You can use the [Vim code coverage information](https://app.codecov.io/gh/vim/vim/tree/master/src) to see the lines of codes that are not yet covered by tests and develop new tests for them.  Note that Vim currently has around 82% code coverage.  This will significantly help in improving the quality of Vim, refactoring of the code base and adding new features without introducing regressions.   This work may not be as fancy as adding new features, but this helps more than adding new features to Vim.
2. Help with triaging the issues reported in [Vim Github issue tracker](https://github.com/vim/vim/issues).
3. Develop test scripts for the issues listed in Github.  This will help in reproducing the issue and adding it to the test suite.
4. Develop fixes for the issues reported in Github.  The [latest todo.txt help](https://github.com/vim/vim/blob/master/runtime/doc/todo.txt) file contains the list of known issues and feature requests.
5. Develop new features listed in the todo.txt file.  The major features that currently need help are virtual text and classes.

You can use the vim-dev mailing list for any development related discussion.

Regards,
Yegappan

**Comment by fritzophrenic** [comment]
> Are there any conventions I should know about to not get made fun of too much :)

Use Vim to develop your code or tests, and turn on the modelines feature. Most of the code (and probably the tests to, I'm not sure) uses a weird mix of tabs and spaces for indentation, but also puts the settings to generate that weird mix right in a modeline near the top or bottom of the file.

Also, for any distributed plugin, syntax, etc. files: check for a maintainer, listed in the file header. They'll usually be the person to actually incorporate the change so you probably need to send it to them (or @ them in a pull request if they're on GitHub).

**Reply by Shane-XB-Qian** [reply]
and to the one there was no maintainer yet: https://github.com/vim/vim/pull/14378
// unless it was wrote and would be maintained by "you" (requester), 
// otherwise it should contact the source repo/author,
// get permission or some discussion, etc.

**Comment by Shane-XB-Qian** [comment]
> uses a weird mix of tabs and spaces for indentation, but also puts the settings to generate that weird mix right in a modeline near the top or bottom of the file.

in Vim codebase, generally it is like this:

```
// to 'C code', put following at every its head:
/* vi:set ts=8 sts=4 sw=4 noet:
*/

" to 'legacy vim script', put following at every its bottom:
" vim: sw=2 sts=2 et

# to 'vim9script', put following at every its bottom:
# vim: sw=2 sts=2 et

- to 'help' doc, put following at every its  bottom:
- note: a white 'space' ahead the 'vim:' here.
 vim:tw=78:ts=8:noet:ft=help:norl:
```

IF you (or someone willing) like, you can start from here, help to add above to its all places.
// it may help to reduce odd indent.

**Reply by Shane-XB-Qian** [reply]
- some other detail guide please check `:h develop.txt`
- some discussion maybe interesting:
-- https://github.com/vim/vim/pull/13294 
-- https://github.com/vim/vim/pull/13301 
-- https://github.com/vim/vim/pull/13358
-- https://github.com/vim/vim/pull/13700#discussion_r1429014637
-- https://github.com/vim/vim/pull/13694#issuecomment-1858902055
-- https://github.com/vim/vim/pull/13642#issuecomment-1849922511

**Reply by Shane-XB-Qian** [reply]
- there perhaps would have formatter https://github.com/vim/vim/pull/13318 but following above advice and guidline maybe more important, https://github.com/vim/vim/pull/16299 should be more wise if unified to use vim code base style.
- there should be also adding proper `test` each time for your change, please try best Not `fix something while break something` : (try best to avoid)
- https://github.com/vim/vim/issues/14338 
- https://github.com/vim/vim/pull/14142
- https://github.com/vim/vim/pull/14186
- https://github.com/vim/vim/pull/14090

**Reply by Shane-XB-Qian** [reply]
- finally (? wish so) please read and follow and insist develop guideline and discussion there: https://github.com/vim/vim/discussions/12736#discussioncomment-6781078 about `what vim is and what vim is not`.
